Integrate PMA

eamsden opened the pull request:
Integrate the PMA work done by @barter-simsum and add logic to stash nouns and cold state in it.


## goals for this PR
We intend to integrate this PR once the following are possible:
- [x] booting to a dojo prompt
- [x] not crashing on sync()
- [x] restoring the pier state and continuing with an interactive dojo prompt

Further issues will be opened for bugfixes and TODOs, but this will allow us to integrate codegen and other pending work against a stable interface.

## archived TODO

**Edward TODO**
- [x] High-level rust bindings
- [x] composable `Persist` trait
- [x] tie PMA into serf
- [x] strip out old PMA and double-jam code
- [x] Change `Hamt` to be pointer-to-stem instead of stem. (It should have been all along, and this facilitates `handle_from_u64`)
- [x] Implement `Persist` for `Noun`
- [x] Implement `Persist` for `Hamt` :sweat: 
- [x] Implement `Persist` for `Batteries`
- [x] Implement `Persist` for `BatteriesList`
- [x] Implement `Persist` for `NounList`
- [x] Implement `Persist` for `Cold`
- [x] Make unifying equality dirty things in the PMA

(after Alex's review)

- [x] Refactor save logic and move into `serf`
- [x] XX comments about where to put warm and hot state so they are not blown away by stack resets
- [x] Remove StemTraversalEntry
- [x] Fix incomplete and out-of-date comments
- [x] Make `Persist for Noun` use `Persist for Atom` implementation for atom case
- [x] Push cell children onto stack in `copy_to_buffer` for `Noun`
- [x] Remove unnecessary `continue`s in HAMT persist loops
- [ ] Make named pointer local vs handle destructuring consistent in `Persist` instances
- [x] Heavy-duty doc-comment on the implementation of the `Preserve` trait and the phase-separation between size-counting and actual copying.
- [x] Heavy-duty doc-comment on the representation of HAMTs

**David TODO**
- [x] add incremented txn id logic to `bt_sync`
- [x] write routine to restore memory maps (became obvious after implementing `_bt_data_cow`
- [x] check return of all mmap calls to not be `MAP_FAILED` - abort if so -- actually should check that the return is the 
fixed addr passed in since mmap makes no guarantee that it will map at that address, but we rely upon that
- [x] complete implementation of bt_state_close() - got neglected
- [ ] c test to verify integrity of btree on a `bt_state_close()` `bt_state_open()` cycle
- [ ] more thorough c tests to validate `_bt_insert` that will exercise deletion coalescing
- [x] `bt_malloc` test - validate modifications to mlist, flist, and possibly btree though that may be adequately covered by `_bt_insert`
- [ ] `bt_free` test. validate modifications to mlist and possibly btree
- [ ] possibly a test that will combine enough `bt_malloc` and `bt_free` calls and follow that up with a call to `bt_sync` to cover a scenario that will drop at least one tree
- [ ] remove CAN_COALESCE macro defn and usage. btree deletion coalescing implemented
- [ ] remove c-side warnings generated during `cargo build`

**Alex TODO**
- [ ] PMA code reviewed (C-side)
- [x] PMA integration code reviewed (Rust-side)

ashelkovnykov commented on rust/ares/src/hamt.rs:
```diff
@@ -481,82 +486,265 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
+            return 0;
+        }
+        let mut bytes: usize = size_of::<Stem<T>>();
+        if pma.contains((*self.0).buffer, (*self.0).size()) {
+            return bytes;
+        };
```
When would this ever happen? If `buffer` is already in the PMA, that implies that it's unchanged from a previous event, and therefore the `bitmap` and `typemap` members should also be unchanged, since they're metadata derived from the `buffer`.

eamsden replied:
The structure of the HAMT is that metadata for the buffer is stored in the *parent'*s buffer, except at the top level where we allocate just one `Stem` struct. A parent can be changed when children aren't, so an altered buffer on the stack could reference an unaltered buffer in the PMA.

ashelkovnykov replied:
> A parent can be changed when children aren't, so an altered buffer on the stack could reference an unaltered buffer in the PMA

Correct, but not at the very top level. Maybe I'm missing and you can think of a counter example, but as I understand it an `insert` call on `Hamt` will always create a new buffer (if you look at the usages of `dest` in `insert`). Therefore, if `pma.contains(self.0, 1)` is `false`, then `pma.contains((*self.0).buffer, (*self.0).size())` is also necessarily `false`.

ashelkovnykov commented on rust/ares/src/hamt.rs:
```diff
@@ -481,82 +486,265 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
+
+            if next_type {
+                // true->stem false->leaf
+                // found another stem
+                traversal[depth + 1] = next_entry.stem;
+
+                if pma.contains(traversal[depth + 1].buffer, traversal[depth + 1].size()) {
```
Shouldn't we check that `&next_entry` isn't already in the PMA first? And if it isn't, don't we need to add `size_of::<Stem<T>>()` to bytes?

eamsden replied:
`next_entry` is temporarily copied from the buffer we are traversing. `Entry` is a union, so the `.stem` here is a union, not struct, field. Entries, not pointers to entries, are elements of the buffers.

ashelkovnykov commented on rust/ares/src/hamt.rs:
```diff
@@ -481,82 +486,265 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
+                // count the buffer for the next stem
+                bytes += traversal[depth + 1].size() * size_of::<Entry<T>>();
+                depth += 1;
+            } else {
+                let mut leaf = next_entry.leaf;
```
Same as above comment, but for `Leaf`

eamsden replied:
Same reply: `Leaf` lives in a `Stem`'s buffer as an `Entry`

ashelkovnykov commented on rust/ares/src/hamt.rs:
```diff
@@ -481,82 +486,265 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
+            traversal[depth].buffer = traversal[depth].buffer.add(1);
+
+            if next_type {
+                // Stem case
```
There's no check here for whether the `Stem` is already in the PMA

eamsden replied:
No, the allocated objects are just the buffers.

ashelkovnykov commented on rust/ares/src/hamt.rs:
```diff
@@ -481,82 +486,265 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
+
+                traversal[depth + 1] = *stem_ptr;
+                depth += 1;
+                continue;
```
Unnecessary `continue`

ashelkovnykov commented on rust/ares/src/hamt.rs:
```diff
@@ -481,82 +486,265 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
+                    leaf_idx += 1;
+                }
+
+                continue;
```
Unnecessary `continue`

ashelkovnykov commented on rust/ares/src/jets/cold.rs:
```diff
@@ -143,6 +200,55 @@ struct BatteriesListMem {
+            let list_mem_ptr = *buffer as *mut BatteriesListMem;
+            copy_nonoverlapping((*dest).0, list_mem_ptr, 1);
+            *buffer = list_mem_ptr.add(1) as *mut u8;
+            (*dest).0 = list_mem_ptr;
+
+            (*(*dest).0).batteries.copy_to_buffer(stack, pma, buffer);
+            dest = &mut (*(*dest).0).next;
```
There's a lot of inconsistency between using the named pointer or `(*dest).0` in the `Persist` implementations in this file. It would be clearer to stick with the named pointer, e.g.:
```
(*dest).0 = list_mem_ptr;

(*list_mem_ptr).batteries.copy_to_buffer(stack, pma, buffer);
dest = &mut (*list_mem_ptr).next;
```

eamsden replied:
Noted, good catch

ashelkovnykov reviewed (commented):
This review is for the Rust side of the PMA integration code only. I'd recommend unit tests for `space_needed` and `copy_to_buffer` for  every single class that implements the `Persist` trait.

eamsden commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+        unsafe { bt_meta_set(self.0, field as usize, val) };
+    }
+
+    pub unsafe fn contains<T>(&self, ptr: *const T, count: usize) -> bool {
+        bt_inbounds(self.0, ptr as *mut c_void) != 0
+            && bt_inbounds(self.0, ptr.add(count) as *mut c_void) != 0
```
> Why...check the end of the allocation as well?
Defensive programming, this is persistence code. Though, if T is in we should simply assert T+count is also in.

> Why.. "in the possible bounds of the PMA"?
Perhaps in contradiction to the above, but there's not an easy way to get the furthest-allocated virtual-memory boundary of the PMA.

eamsden commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+    unsafe fn space_needed(&mut self, stack: &mut NockStack, pma: &PMA) -> usize;
+
+    /// Copy into the provided buffer, which may be assumed to be at least as large as the size
+    /// returned by [space_needed] on the same structure. Return a u64 handle that could be saved
+    /// in metadata
+    unsafe fn copy_to_buffer(&mut self, stack: &mut NockStack, pma: &PMA, buffer: &mut *mut u8);
```
Comment is out of date. Good catch.

Modifying self is consistent with what `Preserve` instances do, and isn't really tricky given that basically all instances of this type are newtype wrappers around pointers. So you're just copying something and then modifying the pointer you were passed in situ.

eamsden commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+    unsafe fn copy_to_buffer(&mut self, stack: &mut NockStack, pma: &PMA, buffer: &mut *mut u8);
+
+    /// Persist an object into the PMA using [space_needed] and [copy_to_buffer], returning
+    /// a [u64] (probably a pointer or tagged pointer) that can be saved into
```
good catch

eamsden commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+}
+
+impl Persist for Atom {
+    unsafe fn space_needed(&mut self, stack: &mut NockStack, pma: &PMA) -> usize {
```
No.

The entire reason for splitting this into two phases is that each phase composes individually. This should almost certainly be called out in detail in a doc comment on the `Persist` trait.

`space_needed` on the instance of a containing type should only ever call `space_needed` on a contained type, never `copy_to_buffer`. Similarly, `copy_to_buffer` should only ever call `copy_to_buffer`, not `space_needed`. And neither should ever call `save_to_pma`.

The whole idea is that you iterate `space_needed` over your entire e.g. `Snapshot` first, and then allocate one buffer in the PMA, then call `copy_to_buffer` over your entire `Snapshot`.

eamsden replied:
In fact, I can't think of a reason `save_to_pma` should ever be re-implemented from the default. So it should probably live in `impl PMA` with a trait-bounded generic, not in the trait itself.

eamsden commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+        if let Ok(indirect) = self.as_indirect() {
+            let count = indirect.raw_size();
+            if !pma.contains(indirect.to_raw_pointer(), count) {
+                if !mark(indirect.as_allocated()) {
```
The definition of `mark` is that it marks if not already marked, and then returns whether the `Allocated` was already marked. Marking is done by setting the high bit of metadata. There should be no forwarding pointers at any point during a `space_needed` call. In fact, an assertion to this effect is probably in order, this being persistence code.

eamsden commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+                    copy_nonoverlapping(indirect.to_raw_pointer(), indirect_buffer_ptr, count);
+                    *buffer = indirect_buffer_ptr.add(count) as *mut u8;
+
+                    indirect.set_forwarding_pointer(indirect_buffer_ptr);
+
+                    *self = IndirectAtom::from_raw_pointer(indirect_buffer_ptr).as_atom();
```
no, self is a mutable reference to the tagged indirect pointer. indirect is the original tagged indirect pointer, and is invalidated once the forwarding pointer is set. Comment forthcoming

ashelkovnykov commented on rust/ares/src/hamt.rs:
```diff
@@ -161,10 +163,11 @@ impl<T: Copy> MutHamt<T> {
 struct Stem<T: Copy> {
     bitmap: u32,
     typemap: u32,
-    buffer: *const Entry<T>,
```
I don't think this should be necessary - change to the buffer implies a new `Stem`, doesn't it?

eamsden replied:
We copy_nonoverlapping the existing buffer first and then `copy_to_buffer` each stem with it, so the mutable pointer is necessary to enable this pattern. I think it is also necessary for unifying equality on keys during lookups

eamsden commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+                    Right(cell) => {
+                        if !pma.contains(cell.to_raw_pointer(), 1) {
+                            if !mark(allocated) {
+                                space += size_of::<CellMemory>();
```
:scream: good catch

eamsden commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+                    *dest = noun;
+                }
+                Right(allocated) => {
+                    if let Some(a) = allocated.forwarding_pointer() {
+                        *dest = a.as_noun();
+                        continue;
+                    }
```
Why would that happen? The only other place forwarding pointers are set is in the noun `Preserve` instance and no live noun should ever contain a forwarding pointer if we're not in the middle of a `preserve` or `copy_to_buffer` call. If the former we're not going to overlap with `copy_to_buffer` and if the latter then the forwarding pointer precisely means that we already copied this noun to the PMA.

eamsden commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+                                continue;
+                            }
+
+                            unmark(allocated);
```
"unmark" is clearing the mark bit in the metadata set by the sizing call, so it does not pollute the noun's metadata in the PMA. Every cell and indirect atom that we actually copy into the PMA will have been marked by `space_needed` and will need to be `unmarked` though @frodwith suggested an optimization to avoid the latter step last CCRS.

eamsden commented on rust/ares/src/serf.rs:
```diff
@@ -82,8 +89,41 @@ impl Context {
+        self.nock_context.stack.reset(0);
+
+        // Since we reset the nockstack the stack-allocated hot state isn't valid anymore
+        self.nock_context.hot = Hot::init(&mut self.nock_context.stack);
```
Yes there should be an `XX` comment here about being able to save the state of the stack and reset to that, thus preserving e.g. the hot state. Also, because allocating the hot state HAMT will generate garbage, we should push, generate the hot state, and copy it into the top frame to keep it there.

But neither of these are expedient now.

eamsden replied:
The above is done

eamsden commented on rust/ares_pma/build.rs:
```diff
@@ -0,0 +1,83 @@
+
+fn main() {
+    let profile = env::var("PROFILE").unwrap();
+    let opt_level = match profile.as_ref() {
```
It should be cc @barter-simsum

eamsden commented on rust/ares/src/hamt.rs:
```diff
@@ -481,82 +486,265 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
-                        traversal_depth -= 1;
-                        continue 'preserve;
+        if stack.is_in_frame(self.0) {
+            let dest_stem = stack.struct_alloc_in_previous_frame(1);
+            copy_nonoverlapping(self.0, dest_stem, 1);
```
copy_nonoverlapping() and rust memory function general take counts in units of the size of the pointed-to type, not bytes. So since our pointer is to Stem, the `1` is "1 `Stem`" not "1 byte"

eamsden commented on rust/ares/src/serf.rs:
```diff
@@ -31,26 +30,34 @@ const FLAG_TRACE: u32 = 1 << 8;
         let newt = Newt::new();
-        let cache = Hamt::<Noun>::new();
+        let cache = Hamt::<Noun>::new(&mut stack);
+        let pma = PMA::open(snap_path).unwrap();
```
I did a big refactor on this code, please re-review.

eamsden commented on rust/ares/src/serf.rs:
```diff
@@ -82,8 +89,41 @@ impl Context {
+        // reset pointers in context to PMA
+        unsafe {
+            self.arvo = (*(snapshot.0)).arvo;
+            self.nock_context.cold = (*(snapshot.0)).cold;
```
ibid

eamsden commented on rust/ares/src/serf.rs:
```diff
@@ -82,8 +89,41 @@ impl Context {
+            (*snapshot_mem_ptr).epoch = self.epoch;
+            (*snapshot_mem_ptr).event_num = self.event_num;
+            (*snapshot_mem_ptr).arvo = self.arvo;
+            (*snapshot_mem_ptr).cold = self.nock_context.cold;
```
ibid

eamsden commented on rust/ares/src/hamt.rs:
```diff
@@ -481,82 +486,265 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
 }
+
+#[derive(Copy, Clone)]
+struct StemTraversalEntry<T: Copy> {
```
I think that's correct

eamsden commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+            let orig_buffer = buffer;
+            self.copy_to_buffer(stack, pma, &mut buffer);
+            assert!(orig_buffer.offset_from(buffer) > 0);
+            assert!(orig_buffer.offset_from(buffer) <= space.try_into().unwrap());
```
> always `=`
No, `<` is not an unallocated write, `>` would be.

eamsden replied:
`space.into()` is probably what should be done here. The point is to have it checked as we are going from an unsigned to signed word of the same bitwidth.

eamsden replied:
**Argument:**
`space_needed` could be conservative (i.e. request more than is necessary) without this being an error condition in the code. I believe it to be exact right now, but if it were to be conservative this should be supported.

**Counterargument:**
Insisting on an exact match would help catch unexpected behaviors, especially missing object copies.

Let me think about this.

eamsden commented:
@ashelkovnykov I believe I've addressed all of your review comments except for the consistency of named pointer handles, which requires some extra consideration on my part that I'm going to push off for a bit.

eamsden commented:
@barter-simsum when you get a chance could you clean up the C-side warnings that are generated when build.rs builds btree.c ?

ashelkovnykov commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,341 @@
+        let mut cold = (*(self.0)).cold;
+        let arvo_space_needed = arvo.space_needed(stack, pma);
+        let cold_space_needed = cold.space_needed(stack, pma);
+        (((size_of::<SnapshotMem>() + 7) >> 3) << 3) + arvo_space_needed + cold_space_needed
```
If:
```
#define MASK 0xfffffff8
```
then:
```
(((x + 7) >> 3) << 3) == (x + 7) & MASK
```
Is there a reason to prefer the former over the latter?

eamsden commented:
Notes from attempts to boot a fakedev with the PMA today:
- I fixed some silly errors in `Persist` and `Preserve` instances.
- I fixed the assertions in `bt_meta_set` and `bt_meta_get`
- It doesn't appear that `ftruncate()` or `truncate()` are actually called anywhere.
- It doesn't appear that `bt_malloc` results in the allocated range being mapped to the file offset returned from the freelist. The first attempt to `save_to_pma` throws SIGBUS from the first copy. Adding `ftruncate` does not fix this, and examining `/proc/<PID>/maps` shows only a whole-PMA linear mapping to the file.

@barter-simsum we should use the pairing session tomorrow to try to get to boot/shutdown/restore for a fakedev.

eamsden commented:
Things to verify
- [ ] `ftruncate()` is properly used to set/extend data.pma
- [ ] `bt_malloc()` properly `mmap`'s the allocated memory range to the allocated disk range
- [ ] `bt_sync()` properly sets protection on cleaned ranges to `PROT_READ` only
- [ ] `bt_free()` properly maps memory range with flags `MAP_FIXED | MAP_ANONYMOUS | MAP_NORESERVE` and protection `PROT_NONE`

@barter-simsum @ashelkovnykov

ashelkovnykov commented on rust/ares/src/hamt.rs:
```diff
@@ -481,82 +497,255 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
+            if next_type {
+                // true->stem false->leaf
+                // found another stem
+                traversal[depth + 1] = next_entry.stem;
+
+                if pma.contains(traversal[depth + 1].buffer, traversal[depth + 1].size()) {
+                    continue;
+                }
```
This is going to needlessly iterate through the buffer of a `Stem` that's already in the PMA. We should be able to skip this as such:
```
if pma.contains(next_entry.stem.buffer, next_entry.stem.size()) {
    continue;
}

depth += 1;
traversal[depth] = next_entry.stem;
// count the buffer for the next stem
bytes += traversal[depth].size() * size_of::<Entry<T>>();
```

eamsden replied:
Not so. This is checking whether the buffer of a *child* `Stem` to the stem at `depth` is already in the PMA, in which case we will skip preserving it.

ashelkovnykov replied:
Correct, but if the buffer of the child `Stem` is already in the PMA, it stands to reason that the buffers of any children of *that* `Stem` are already in the PMA as well. However, we're going to traverse and check for each of those children, because we've increased `depth` by `1`.

eamsden replied:
We haven't though. Look a few lines below. We `continue` in this `if` but increment depth below the `if`. We do write the child node into the traversal stack at `depth + 1` but that does not mutate `depth`.

ashelkovnykov replied:
Bleh - you're right. Somehow I conflated "assignment to `traversal[depth + 1]`" with "`depth` incremented". I think I got confused because `traversal` changes, even though `depth` doesn't, in this case.

ashelkovnykov commented on rust/ares/src/hamt.rs:
```diff
@@ -481,82 +497,255 @@ impl<T: Copy + Preserve> Preserve for Hamt<T> {
+        if pma.contains(self.0, 1) {
+            return 0;
+        }
+        let mut bytes: usize = size_of::<Stem<T>>();
```
I believe we're currently forgetting to add the size of the root `Stem`s buffer to `bytes`.

eamsden replied:
No. If the root stem is already in the PMA then we don't need to do any copying. Otherwise we initialize the `bytes` counter to the size of one `Stem`: the root stem.

ashelkovnykov replied:
But if the buffer of the root `Stem` isn't in the PMA, we never increase `bytes` by the size of the buffer.

eamsden replied:
I just realized you were right about this and fixed it.

eamsden commented on rust/ares/src/serf.rs:
```diff
@@ -28,47 +29,157 @@ crate::gdb!();
+    ) -> Context {
+        let mut pma = PMA::open(snap_path).expect("serf: pma open failed");
+
+        let snapshot_version = pma.meta_get(BTMetaField::SnapshotVersion as usize);
```
This is not the case. Roots is just a place to put handles to things stored in the PMA. But yes, the names should be consistent, which also requires changing the C API.

cc @barter-simsum

ashelkovnykov commented on rust/ares/src/persist.rs:
```diff
@@ -265,6 +265,7 @@ impl Persist for Noun {
             }
         }
         *buffer = buffer_u64 as *mut u8;
+        stack.frame_pop();
```
How did we both miss this...

eamsden replied:
Dunno but I caught it when `flip_top_frame`'s assertion that it was, in fact, in the top frame failed.

ashelkovnykov commented on rust/ares/src/serf.rs:
```diff
@@ -79,22 +188,37 @@ impl Context {
+    pub unsafe fn preserve_event_update_leftovers(&mut self) {
+        let stack = &mut self.nock_context.stack;
+        stack.preserve(&mut self.arvo);
+        stack.preserve(&mut self.nock_context.cold);
```
Why is preserving cold necessary - shouldn't it be in the PMA by this point? And any dangling forward pointers in other objects resolved by calling preserve on those objects?

eamsden replied:
It's not necessary, these two lines can be removed. I was just cut-and-pasta-ing the stack flip from the bottom of the serf loop

ashelkovnykov commented on rust/ares/src/mem.rs:
```diff
@@ -51,6 +50,7 @@ pub struct NockStack {
     alloc_pointer: *mut u64,
     /** MMap which must be kept alive as long as this NockStack is */
     memory: MmapMut,
+    /** PMA from which we will copy into the NockStack */
```
Incoming member that's not yet present?

ashelkovnykov commented on rust/ares/src/mem.rs:
```diff
@@ -106,6 +143,26 @@ impl NockStack {
         self.frame_pointer
     }
 
+    /** Current stack pointer of this NockStack */
+    pub fn get_stack_pointer(&self) -> *const u64 {
+        self.stack_pointer
+    }
+
+    /** Current alloc pointer of this NockStack */
+    pub fn get_alloc_pointer(&self) -> *const u64 {
+        self.alloc_pointer
+    }
+
+    /** Start of the memory range for this NockStack */
+    pub fn get_start(&self) -> *const u64 {
+        self.start
+    }
+
+    /** End of the memory range for this NockStack */
+    pub fn get_size(&self) -> usize {
+        self.size
+    }
+
```
I keep adding these for debugging, then removing them, too. Finally someone had the wherewithal to actually commit them...

eamsden replied:
unifying equality moved to its own module, so we need public interfaces to this stuff.

ashelkovnykov commented on rust/ares/src/persist.rs:
```diff
@@ -0,0 +1,315 @@
+    }
+}
+
+/** Mask to mask out pointer bits not aligned with a BT_PAGESIZE page */
+const BT_PAGEBITS_MASK_OUT: u64 = !((1 << BT_PAGEBITS) - 1);
+
+// round an address down to a page boundary
+fn bt_page_round_down<T>(ptr: *mut T) -> *mut c_void {
+    ((ptr as u64) & BT_PAGEBITS_MASK_OUT) as *mut c_void
+}
+
+// round an address up to a page boundary
+fn bt_page_round_up<T>(ptr: *mut T) -> *mut c_void {
+    (((ptr as u64) + (BT_PAGESIZE as u64) - 1) & BT_PAGEBITS_MASK_OUT) as *mut c_void
+}
```
Anyway to have this in `btree.c` and call it from Rust? I don't like the idea of duplicating functions / constants between the two languages.

eamsden replied:
These functions are unused in `btree.c`, which exposes its interface in page offsets, not byte-level pointers.

eamsden replied:
I agree with your opinion, but the functions are not duplicated (they are only rust-side) and the constants are imported via rust-bindgen from `btree.h`

eamsden commented:
@ashelkovnykov @joemfb the plan is to merge this once it is capable of
- [x] booting to a dojo prompt
- [x] not crashing on sync()
- [ ] restoring the pier state and continuing with an interactive dojo prompt

Please review with an eye toward this merge. Bugs, TODOs, and cleanup which are not on the path to merging the above functionality with a CI greencheck are to be filed as issues and worked on in separate PRs.

I believe the interface for this is sufficiently stabilized that merging it will enable me to focus on integrating other components, most notably codegen.

ashelkovnykov reviewed (approved):
I understand 100% of the Rust changes and 15% of the C changes - shipping a review to get this out of monolithic status.
